### PR TITLE
docs: add a testing cheat code file

### PR DESCRIPTION
**Pull request by od-hunter**
Resolves https://github.com/dojoengine/book/issues/300

### CODE REVIEW

**Review by glihm — CHANGES_REQUESTED** [review]
Thanks for the work here.

Could you please add more realistic examples that actually work, using the correct `use` and instructions to actually access those functions from the corelib?

No problem at all with AI generated content, please make sure it's actually accurate.

Please don't hesitate if you have any question.

**Review by glihm — CHANGES_REQUESTED** [review]
> Could you please add more realistic examples that actually work, using the correct use and instructions to actually access those functions from the corelib?

Thanks for the work here, your examples are still using invalid Cairo Code. Could you please set Cairo examples that can be compiled to ensure the user can correctly test it? Thanks!

**Comment by od-hunter** [comment]
@glihm please review, I’ve made necessary corrections

**Review by glihm — COMMENTED** [review]
> @glihm please review, I’ve made necessary corrections


Hum, I don't think you have the will to learn here, mostly copy-pasting wrong part of the code and some AI text and no actual understanding of the code.

Here's an other link to help you going through the implementation again: https://github.com/starkware-libs/cairo/blob/72d1d2f50e8c2c13d5414cc4365683299d8f3c8b/crates/cairo-lang-starknet/cairo_level_tests/l2_to_l1_messages.cairo#L69.

Would appreciate if the next iteration actually contain some changes that are relevant for the reader of the book. Otherwise, I'll close and un-assigned.

**Comment by od-hunter** [comment]
Sir please I need your telegram I have questions to ask you

**Comment by glihm** [comment]
> Sir please I need your telegram I have questions to ask you

No problem at all, you can come in Dojo discord any time, happy to answer any question. 🫡

File: pnpm-lock.yaml
Comment (glihm) [inline pnpm-lock.yaml:1]: Could you revert this file please?

File: package-lock.json
Comment (glihm) [inline package-lock.json:1]: This file should be reverted too. No changes expected. 👍

File: docs/pages/framework/testing-cheat-codes.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+The Cairo Testing Cheat Codes allow you to set and manipulate various execution context variables, such as block number, caller address, contract address, to test your contracts in different scenarios. In this guide, we will explore each cheat code in detail, providing explanations and examples to help you understand how to use them effectively in your tests.
+
+
+1. `set_block_number`
+
+This cheat code helps you set the current block number to the specified value, allowing you to simulate different block heights for testing purposes.
+
+- Testing a contract's behavior at a specific block height, such as checking if a certain function is only callable after a certain block number.
+- Simulating a fork or replay attack to test a contract's resilience.
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:11]:
```suggestion
```

File: docs/pages/framework/testing-cheat-codes.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
+# Cairo Testing Cheat Codes: A Comprehensive Guide
+
+The Cairo Testing Cheat Codes allow you to set and manipulate various execution context variables, such as block number, caller address, contract address, to test your contracts in different scenarios. In this guide, we will explore each cheat code in detail, providing explanations and examples to help you understand how to use them effectively in your tests.
+
+
+1. `set_block_number`
+
+This cheat code helps you set the current block number to the specified value, allowing you to simulate different block heights for testing purposes.
+
+- Testing a contract's behavior at a specific block height, such as checking if a certain function is only callable after a certain block number.
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:10]:
```suggestion
This one is helpful to test a contract's behavior at a specific block height, such as checking if a certain function is only callable after a certain block number.
```

File: docs/pages/framework/testing-cheat-codes.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+- Simulating a fork or replay attack to test a contract's resilience.
+
+```
+use starknet::{testing, get_block_number};
+
+testing::set_block_number(1234567);
+
+get_block_nunmber(); // returns 1234567
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:18]:
```suggestion
assert!(get_block_number() == 1234567, 'bad block number');
```

File: docs/pages/framework/testing-cheat-codes.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+const USER_ONE: felt252 = 'USER1';
+testing::set_caller_address(USER_ONE.try_into().unwrap()); // Sets user to USER_ONE
+
+get_caller_address(); // returns USER_ONE.try_into().unwrap()
+
+```
+
+3. `set_contract_address`
+
+This cheat code helps you set the contract address to the provided value, allowing you to test contract deployment and interactions. You can apply this code when:
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:41]:
```suggestion
This cheat code helps you set the contract address to the provided value, allowing you to test contract deployment and interactions.
```

File: docs/pages/framework/testing-cheat-codes.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+- Deploying a contract to a specific address for testing purposes.
+- Testing a contract's behavior when deployed to a different address.
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:44]:
```suggestion
It is important to note that any test function is considered a contract, which by default uses the `0` address. Using `set_contract_address` allows you to mock the current address of the testing function, making it useful to call other contract that may use `get_caller_address`.
```

File: docs/pages/framework/testing-cheat-codes.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+```
+use core::traits::{TryInto, Into};
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:47]:
```suggestion
```
Shouldn't be necessary.

File: docs/pages/framework/testing-cheat-codes.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+use starknet::{testing, get_caller_address};
+
+const USER_ONE: felt252 = 'USER1';
+testing::set_caller_address(USER_ONE.try_into().unwrap()); // Sets user to USER_ONE
+
+get_caller_address(); // returns USER_ONE.try_into().unwrap()
+
```
Comment (glihm) [inline docs/pages/framework/testing-cheat-codes.md:36]:
```suggestion
use starknet::{testing, get_caller_address, contract_address_const};

let user_one = contract_address_const::<'user1'>();
testing::set_caller_address(user_one);
assert(get_caller_address() == user_one, 'bad caller';
```

**Review by glihm — CHANGES_REQUESTED** [review]
Thanks for the work done here @od-hunter.

I've corrected the two first, to give you some examples of working cairo.

Could you please review all the others to ensure they are compilable and working?

I suggest you to do the following to test them:

1. `scarb new my_proj` to create a new project.
2. Set the `Scarb.toml` to build starknet contract, [full example available here](https://github.com/software-mansion/scarb/tree/main/examples/starknet_hello_world).
3. Write one function for each cheat code with the same style I've given in this review. We prefer using assert instead of comments.
4. Finally, run `scarb test` to verify all your tests are green, and then you can just copy paste the code in each cheat code.

Thank you!

**Pull request closed without merging**